src/include/simeng/config/Expectations.hh

```diff
@@ -0,0 +1,361 @@
+#pragma once
+
+#include <math.h>
+
+#include <algorithm>
+#include <cassert>
+#include <climits>
+#include <fstream>
+#include <iostream>
+#include <queue>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <variant>
+
+#include "simeng/config/yaml/ryml.hh"
+
+namespace simeng {
+namespace config {
+
+/** An enum containing all supported data types that can be expected of a config
+ * option. */
+enum ExpectedType { Bool, Double, Float, Integer, String, UInteger, Valueless };
+
+/** String to represent a wildcard node key. */
+const std::string wildcard = "*";
+
+/** A class to hold the expectations of a specific config option. Each class
+ * is considered to be one node of a tree-like structure which maps onto the
+ * hierarchical YAML structure of the passed/generated config file. Each node
+ * can contain any number of children, each of which is another instance of the
+ * `expectationNode` class for another config option. The expectation placed on
+ * each config option can be defined as a type, a set of values to which it must
+ * belong, and a set of bounds it must lie between. A default value is also
+ * expected for the sake of default construction and generation of default
+ * config files. The values of such expectations are held within a
+ * `std::variant` which can hold one of the expected data types equivalent to
+ * that held in the `ExpectedType` enum. */
+class expectationNode {
+  using dataTypeVariant =
+      std::variant<bool, double, float, int64_t, std::string, uint64_t>;
+
+ public:
+  expectationNode(){};
+
+  ~expectationNode(){};
+
+  /** A templated function to allow for the creation of an `expectationNode`
+   * instance. The instance created is one with a value and a key. A key,
+   * default value, type, and a flag denoting whether the node is optional are
+   * provided to the underlying constructor. */
+  template <typename T, bool optional = false>
+  expectationNode create(std::string key, T defaultValue) {
```

Comment:
Not quite sure what the original question was posing but think I've figured it out. I've added this static_assert to ensure the passed type matches one supported.

Reply:
Proceeding from [this](https://github.com/UoB-HPC/SimEng/pull/307/files#r1202469737) comment.

The statement `ExpectedType type = static_cast<ExpectedType>(defValVariant.index())` will never resolve to `ExpectedType::Valueless` because the maximum index returned by `defValVariant.index()` will be 5. If a type other than the types in the type definition of `dataTypeVariant` is passed the variant will complain during compilation.

Also given my previous [comment](https://github.com/UoB-HPC/SimEng/pull/307/files#r1200477496) adding the SFINAE template will also make sure no types other can be passed; hence ensuring `ExpectedType::Valueless` is never returned.

Reply:
Resolving "ExpectedType type = static_cast<ExpectedType>(defValVariant.index()) will never resolve". Keeping `ExpectedType::Valueless` as it's needed to ensure a valid type of the parent ExpectationNodes and is used within the `validateConfigNode` call.